File: backend/python/test_harness.py
"""
Test harness — runs the candidate's fetch_all against the test server and
validates correctness.

Start the server first (from the typescript directory):
    cd ../typescript && npm run server

Usage:  uv run test_harness.py [path/to/solution.py]
"""

import asyncio
import importlib.util
import json
import os
import sys
import time
import urllib.request
from pathlib import Path

solution_path = Path(sys.argv[1] if len(sys.argv) > 1 else "solution.py").resolve()
spec = importlib.util.spec_from_file_location("solution", solution_path)
assert spec and spec.loader, f"Could not load {solution_path}"
_mod = importlib.util.module_from_spec(spec)
spec.loader.exec_module(_mod)
Success = _mod.Success  # type: ignore[attr-defined]
Failure = _mod.Failure  # type: ignore[attr-defined]
fetch_all = _mod.fetch_all  # type: ignore[attr-defined]

SERVER = os.environ.get("SERVER_URL", "http://localhost:3000")


def _get(path: str):
    with urllib.request.urlopen(f"{SERVER}{path}") as resp:
        return json.loads(resp.read())


def _post(path: str):
    req = urllib.request.Request(f"{SERVER}{path}", method="POST")
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read())


async def main() -> None:
    # ------------------------------------------------------------------
    # Setup
    # ------------------------------------------------------------------
    try:
        urls: list[str] = _get("/urls")
    except Exception as exc:
        print(f"Could not reach server at {SERVER}: {exc}")
        sys.exit(1)

    print(f"Solution: {solution_path}")
    print(f"Fetched {len(urls)} URLs from server\n")
    _post("/reset")

    # ------------------------------------------------------------------
    # Run candidate solution
    # ------------------------------------------------------------------
    start = time.perf_counter()
    try:
        results = await fetch_all(
            urls,
            max_concurrent=10,
            max_requests_per_second=15,
        )
    except Exception as exc:
        print(f"FAIL — fetch_all raised an exception: {exc}")
        sys.exit(1)
    elapsed = time.perf_counter() - start

    # ------------------------------------------------------------------
    # Validate return value
    # ------------------------------------------------------------------
    if not isinstance(results, list):
        print(
            f"FAIL — fetch_all must return a list, got {type(results).__name__}: {results!r}"
        )
        sys.exit(1)

    # ------------------------------------------------------------------
    # Validate result shape
    # ------------------------------------------------------------------
    invalid = [
        (i, r)
        for i, r in enumerate(results)
        if not (isinstance(r, Success) or isinstance(r, Failure))
    ]
    if invalid:
        print(
            f"FAIL — {len(invalid)} result(s) have an invalid type."
            " Each result must be a Success or Failure instance."
        )
        for i, r in invalid[:5]:
            print(f"  [{i}]: {r!r}")
        if len(invalid) > 5:
            print(f"  ... and {len(invalid) - 5} more")
        sys.exit(1)

    # ------------------------------------------------------------------
    # Checks
    # ------------------------------------------------------------------
    passed = 0
    failed = 0

    def check(name: str, ok: bool, detail: str = "") -> None:
        nonlocal passed, failed
        if ok:
            print(f"  PASS  {name}")
            passed += 1
        else:
            msg = f"  FAIL  {name}"
            if detail:
                msg += f" — {detail}"
            print(msg)
            failed += 1

    def get(i: int):
        """Safe index access — returns None for out-of-bounds."""
        return results[i] if i < len(results) else None

    def describe(r) -> str:
        """Human-readable description of a result or missing value."""
        if r is None:
            return "missing"
        if isinstance(r, Failure):
            return r.error
        if isinstance(r, Success):
            return "success"
        return repr(r)

    print("Results:")

    # 1. Coverage
    check(
        "All URLs covered",
        len(results) == len(urls),
        f"expected {len(urls)}, got {len(results)}",
    )

    # 2. Ordering preserved
    ordered = all(
        getattr(get(i), "url", None) == urls[i] for i in range(len(urls))
    )
    check("Ordering preserved", ordered)

    # 3. Transient errors recovered (IDs 90-94 → should end up as Success)
    transient_ids = range(90, 95)
    transient_ok = all(isinstance(get(i), Success) for i in transient_ids)
    bad_t = [
        f"item {i}: {describe(get(i))}"
        for i in transient_ids
        if not isinstance(get(i), Success)
    ]
    check("Transient 5xx errors retried and recovered", transient_ok, ", ".join(bad_t))

    # 4. Permanent errors captured as Failure (IDs 95-99 → Failure, not thrown)
    perm_ids = range(95, 100)
    perm_ok = all(isinstance(get(i), Failure) for i in perm_ids)
    bad_p = [
        f"item {i}: {describe(get(i))}"
        for i in perm_ids
        if not isinstance(get(i), Failure)
    ]
    check("Permanent errors captured (not thrown)", perm_ok, ", ".join(bad_p))

    # 5. Fast items (0-79)
    fast_ok = len(results) >= 80 and all(isinstance(get(i), Success) for i in range(80))
    check("Fast items (0-79) succeeded", fast_ok)

    # 6. Slow items (80-89)
    slow_ok = len(results) >= 90 and all(isinstance(get(i), Success) for i in range(80, 90))
    check("Slow items (80-89) succeeded", slow_ok)

    # 7. Concurrency — server tracked peak in-flight requests
    stats = _get("/stats")
    peak = stats.get("peakConcurrency", 0)
    check(
        "Used concurrency",
        peak > 1,
        f"peak was {peak} (sequential)",
    )
    check(
        "Respected concurrency limit",
        peak <= 10,
        f"peak was {peak}, limit is 10",
    )

    # 8. Rate limiting — server tracked peak requests in any 1-second window
    peak_rps = stats.get("peakWindowRequests", 0)
    check(
        "Respected rate limit",
        peak_rps <= 15,
        f"peak was {peak_rps} req/s, limit is 15",
    )

    # 9. Server confirms retries actually happened (IDs 90-94 need >= 2 attempts)
    attempt_counts = stats.get("attemptCounts", {})
    retried_on_server = all(
        attempt_counts.get(str(i), 0) >= 2 for i in range(90, 95)
    )
    bad_r = [
        f"item {i}: {attempt_counts.get(str(i), 0)} attempt(s)"
        for i in range(90, 95)
        if attempt_counts.get(str(i), 0) < 2
    ]
    check("Transient items retried (server-side)", retried_on_server, ", ".join(bad_r))

    # ------------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------------
    print(f"\n{passed} passed, {failed} failed — {elapsed:.2f}s elapsed")
    print(f"\nServer stats: {json.dumps(stats, indent=2)}")

    sys.exit(1 if failed > 0 else 0)


if __name__ == "__main__":
    asyncio.run(main())

File: backend/typescript/test-harness.ts
/**
 * Test harness — runs the candidate's fetchAll against the test server and
 * validates correctness.  Start the server first:  npm run server
 *
 * Usage:  npx tsx test-harness.ts [path/to/solution.ts]
 */

import { resolve } from "node:path";

const solutionPath = resolve(process.argv[2] || "./solution.ts");
const { fetchAll } = await import(solutionPath);

type Result =
  | { url: string; body: string }
  | { url: string; error: string };

const SERVER = process.env.SERVER_URL || "http://localhost:3000";

function isSuccess(r: Result | undefined): r is { url: string; body: string } {
  return r != null && typeof r === "object" && "body" in r && !("error" in r);
}

function isFailure(r: Result | undefined): r is { url: string; error: string } {
  return r != null && typeof r === "object" && "error" in r;
}

function describeResult(r: Result | undefined): string {
  if (r == null) return "undefined";
  if ("error" in r) return r.error;
  if ("body" in r) return "success";
  return JSON.stringify(r);
}

async function main(): Promise<void> {
  console.log(`Solution: ${solutionPath}\n`);
  // ------------------------------------------------------------------
  // Setup
  // ------------------------------------------------------------------
  const urlsRes = await fetch(`${SERVER}/urls`);
  if (!urlsRes.ok) {
    console.error("Could not reach server at", SERVER);
    process.exit(1);
  }
  const urls: string[] = await urlsRes.json() as string[];
  console.log(`Fetched ${urls.length} URLs from server\n`);

  await fetch(`${SERVER}/reset`, { method: "POST" });

  // ------------------------------------------------------------------
  // Run candidate solution
  // ------------------------------------------------------------------
  const start = performance.now();
  let results: Result[];
  try {
    results = await fetchAll(urls, {
      maxConcurrent: 10,
      maxRequestsPerSecond: 15,
    });
  } catch (err) {
    console.error("FAIL — fetchAll threw an exception:", err);
    process.exit(1);
  }
  const elapsed = ((performance.now() - start) / 1000).toFixed(2);

  // ------------------------------------------------------------------
  // Validate return value
  // ------------------------------------------------------------------
  if (!Array.isArray(results)) {
    console.error(
      `FAIL — fetchAll must return an array, got ${typeof results}: ${JSON.stringify(results)}`,
    );
    process.exit(1);
  }

  // ------------------------------------------------------------------
  // Validate result shape
  // ------------------------------------------------------------------
  function isValidResult(r: unknown): r is Result {
    if (r == null || typeof r !== "object") return false;
    if (!("url" in r) || typeof (r as any).url !== "string") return false;
    // Must be either a success (has body) or a failure (has error)
    if ("body" in r && typeof (r as any).body === "string") return true;
    if ("error" in r && typeof (r as any).error === "string") return true;
    return false;
  }

  const invalid = results
    .map((r, i) => [i, r] as const)
    .filter(([, r]) => !isValidResult(r));

  if (invalid.length > 0) {
    console.error(
      `FAIL — ${invalid.length} result(s) have an invalid shape.` +
      ` Each result must have { url, status, body } or { url, error }.`,
    );
    for (const [i, r] of invalid.slice(0, 5)) {
      console.error(`  [${i}]:`, JSON.stringify(r));
    }
    if (invalid.length > 5) {
      console.error(`  ... and ${invalid.length - 5} more`);
    }
    process.exit(1);
  }

  // ------------------------------------------------------------------
  // Checks
  // ------------------------------------------------------------------
  let passed = 0;
  let failed = 0;

  function check(name: string, ok: boolean, detail?: string): void {
    if (ok) {
      console.log(`  PASS  ${name}`);
      passed++;
    } else {
      console.log(`  FAIL  ${name}${detail ? ` — ${detail}` : ""}`);
      failed++;
    }
  }

  console.log("Results:");

  // 1. Coverage
  check(
    "All URLs covered",
    results.length === urls.length,
    `expected ${urls.length}, got ${results.length}`,
  );

  // 2. Ordering preserved
  const ordered = results.every((r, i) => r?.url === urls[i]);
  check("Ordering preserved", ordered);

  // 3. Transient errors recovered (IDs 90-94 → should end up as Success)
  const transientOk = [90, 91, 92, 93, 94].every(
    (id) => isSuccess(results[id]),
  );
  check(
    "Transient 5xx errors retried and recovered",
    transientOk,
    [90, 91, 92, 93, 94]
      .filter((id) => !isSuccess(results[id]))
      .map((id) => `item ${id}: ${describeResult(results[id])}`)
      .join(", "),
  );

  // 4. Permanent errors captured as Failure (IDs 95-99 → Failure, not thrown)
  const permanentOk = [95, 96, 97, 98, 99].every(
    (id) => isFailure(results[id]),
  );
  check(
    "Permanent errors captured (not thrown)",
    permanentOk,
    [95, 96, 97, 98, 99]
      .filter((id) => !isFailure(results[id]))
      .map((id) => `item ${id}: ${describeResult(results[id])}`)
      .join(", "),
  );

  // 5. Fast items (0-79)
  const fastOk =
    results.length >= 80 &&
    results.slice(0, 80).every((r) => isSuccess(r));
  check("Fast items (0-79) succeeded", fastOk);

  // 6. Slow items (80-89)
  const slowOk =
    results.length >= 90 &&
    results.slice(80, 90).every((r) => isSuccess(r));
  check("Slow items (80-89) succeeded", slowOk);

  // 7. Concurrency — server tracked peak in-flight requests
  const statsRes = await fetch(`${SERVER}/stats`);
  const stats = (await statsRes.json()) as {
    rateLimit: number;
    currentWindowRequests: number;
    peakConcurrency: number;
    peakWindowRequests: number;
    attemptCounts: Record<string, number>;
  };
  check(
    "Used concurrency",
    stats.peakConcurrency > 1,
    `peak was ${stats.peakConcurrency} (sequential)`,
  );
  check(
    "Respected concurrency limit",
    stats.peakConcurrency <= 10,
    `peak was ${stats.peakConcurrency}, limit is 10`,
  );

  // 8. Rate limiting — server tracked peak requests in any 1-second window
  check(
    "Respected rate limit",
    stats.peakWindowRequests <= 15,
    `peak was ${stats.peakWindowRequests} req/s, limit is 15`,
  );

  // 9. Server confirms retries actually happened (IDs 90-94 need >= 2 attempts)
  const retriedOnServer = [90, 91, 92, 93, 94].every(
    (id) => (stats.attemptCounts[id] ?? 0) >= 2,
  );
  check(
    "Transient items retried (server-side)",
    retriedOnServer,
    [90, 91, 92, 93, 94]
      .filter((id) => (stats.attemptCounts[id] ?? 0) < 2)
      .map((id) => `item ${id}: ${stats.attemptCounts[id] ?? 0} attempt(s)`)
      .join(", "),
  );

  // ------------------------------------------------------------------
  // Summary
  // ------------------------------------------------------------------
  console.log(`\n${passed} passed, ${failed} failed — ${elapsed}s elapsed`);
  console.log("\nServer stats:", JSON.stringify(stats, null, 2));

  process.exit(failed > 0 ? 1 : 0);
}

main();
